adapters/coap-vertx/src/main/java/org/eclipse/hono/adapter/coap/vertx/VertxBasedCoapAdapter.java

```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 1.0 which is available at
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ */
+
+package org.eclipse.hono.adapter.coap.vertx;
+
+import java.util.List;
+
+import org.eclipse.californium.core.CoapServer;
+import org.eclipse.californium.core.coap.CoAP.ResponseCode;
+import org.eclipse.californium.core.server.resources.CoapExchange;
+import org.eclipse.hono.adapter.coap.AbstractVertxBasedCoapAdapter;
+import org.eclipse.hono.adapter.coap.CoapAdapterProperties;
+import org.eclipse.hono.service.auth.device.Device;
+import org.eclipse.hono.util.Constants;
+import org.eclipse.hono.util.EventConstants;
+import org.eclipse.hono.util.TelemetryConstants;
+
+import io.vertx.core.Context;
+import io.vertx.core.Handler;
+
+/**
+ * A Vert.x based Hono protocol adapter for accessing Hono's Telemetry &amp; Event API using COAP.
+ */
+public final class VertxBasedCoapAdapter extends AbstractVertxBasedCoapAdapter<CoapAdapterProperties> {
+
+    /**
+     * {@inheritDoc}
+     * 
+     * @return {@link Constants#PROTOCOL_ADAPTER_TYPE_COAP}
+     */
+    @Override
+    protected String getTypeName() {
+        return Constants.PROTOCOL_ADAPTER_TYPE_COAP;
+    }
+
+    /**
+     * Get extended device.
+     * 
+     * @param exchange coap exchange with URI.
+     * @param identityLevel path level to extract URI identity. e.g. for "/telemetry/tenant/authid" the level is 1.
+     * @param handler handler for determined extended device
+     */
+    public void getExtendedDevice(final CoapExchange exchange, final int identityLevel,
+            final Handler<ExtendedDevice> handler) {
+        if (identityLevel > 0) {
+            final List<String> path = exchange.getRequestOptions().getUriPath();
+            if (path.size() == identityLevel + 2) {
+                final Device device = new Device(path.get(identityLevel), path.get(identityLevel + 1));
+                final ExtendedDevice extendedDevice = new ExtendedDevice(device, device);
+                if (extendedDevice != null) {
+                    log.debug("use {}", extendedDevice);
+                    handler.handle(extendedDevice);
+                    return;
+                }
+            }
+        }
+        exchange.respond(ResponseCode.INTERNAL_SERVER_ERROR);
```

Comment:
we usually try to return a more specific error, e.g. if an unauthenticated device's request URI does not contain a tenant and device ID, the HTTP adapter returns a 404